### PR TITLE
[CPP Graph] KV-Update Optimization

### PR DESCRIPTION
**Pull request by DDEle**
## Type of Change: feature
API not changed

## Description
- The FLASH_ATTN_KV_UPDATE op now uses JIT implementations for first token inference
- Update jblas to 21757a4

## Expected Behavior & Potential Risk
N/A

## How has this PR been tested?
<details>
<summary>Interference result:</summary>

```bash
set -ex
#main branch
rm -rf bin && cmake .. -GNinja -DNE_BUILD_TESTS=ON -DNE_PROFILING=ON -DCMAKE_BUILD_TYPE=Release && ninja run_gptj && env ENGINE_PROFILING=1 numactl -m 1 -C 56-111 bin/run_gptj -m /home_local/dingyi/data/gpt-j-6B-pr347-q4j-sym-int8-fp32-g128.bin --seed 1234 -t 48 -b 2048 -c 2048 -n 1 -p "$LUOYU_PROMPT"
rm -rf bin && cmake .. -GNinja -DNE_BUILD_TESTS=ON -DNE_PROFILING=ON -DCMAKE_BUILD_TYPE=Release && ninja run_gptj && env ENGINE_PROFILING=1 numactl -m 1 -C 56-111 bin/run_gptj -m /home_local/dingyi/data/gpt-j-6B-pr347-q4j-sym-int8-fp32-g128.bin --seed 1234 -t 56 -b 2048 -c 2048 -n 1 -p "$LUOYU_PROMPT"
rm -rf bin && cmake .. -GNinja -DNE_BUILD_TESTS=OFF -DNE_PROFILING=OFF -DCMAKE_BUILD_TYPE=Release && ninja run_gptj && env --unset ENGINE_PROFILING numactl -m 1 -C 56-111 bin/run_gptj -m /home_local/dingyi/data/gpt-j-6B-pr347-q4j-sym-int8-fp32-g128.bin --seed 1234 -t 56 -b 2048 -c 2048 -n 32 -p "$GIRL_PROMPT"
#llama has different tokenizer and can only process a slightly shorter prompt
rm -rf bin && cmake .. -GNinja -DNE_BUILD_TESTS=ON -DNE_PROFILING=ON -DCMAKE_BUILD_TYPE=Release && ninja run_llama && env ENGINE_PROFILING=1 numactl -m 1 -C 56-111 bin/run_llama -m /home_local/dingyi/data/llama-7b-hf-pr347-q4j-sym-int8-fp32-g128.bin --seed 1234 -t 56 -b 2048 -c 2048 -n 1 -p "$(echo "$LUOYU_PROMPT" | cut -d' ' -f 1-1500)"
rm -rf bin && cmake .. -GNinja -DNE_BUILD_TESTS=OFF -DNE_PROFILING=OFF -DCMAKE_BUILD_TYPE=Release && ninja run_llama && env --unset ENGINE_PROFILING numactl -m 1 -C 56-111 bin/run_llama -m /home_local/dingyi/data/llama-7b-hf-pr347-q4j-sym-int8-fp32-g128.bin --seed 1234 -t 56 -b 2048 -c 2048 -n 32 -p "$GIRL_PROMPT"

#this branch
rm -rf bin && cmake .. -GNinja -DNE_BUILD_TESTS=ON -DNE_PROFILING=ON -DCMAKE_BUILD_TYPE=Release && ninja run_gptj && env ENGINE_PROFILING=1 numactl -m 1 -C 56-111 bin/run_gptj -m /home_local/dingyi/data/gpt-j-6B-pr347-q4j-sym-int8-fp32-g128.bin --seed 1234 -t 48 -b 2048 -c 2048 -n 1 -p "$LUOYU_PROMPT"
rm -rf bin && cmake .. -GNinja -DNE_BUILD_TESTS=ON -DNE_PROFILING=ON -DCMAKE_BUILD_TYPE=Release && ninja run_gptj && env ENGINE_PROFILING=1 numactl -m 1 -C 56-111 bin/run_gptj -m /home_local/dingyi/data/gpt-j-6B-pr347-q4j-sym-int8-fp32-g128.bin --seed 1234 -t 56 -b 2048 -c 2048 -n 1 -p "$LUOYU_PROMPT"
rm -rf bin && cmake .. -GNinja -DNE_BUILD_TESTS=OFF -DNE_PROFILING=OFF -DCMAKE_BUILD_TYPE=Release && ninja run_gptj && env --unset ENGINE_PROFILING numactl -m 1 -C 56-111 bin/run_gptj -m /home_local/dingyi/data/gpt-j-6B-pr347-q4j-sym-int8-fp32-g128.bin --seed 1234 -t 56 -b 2048 -c 2048 -n 32 -p "$GIRL_PROMPT"
#llama has different tokenizer and can only process a slightly shorter prompt
rm -rf bin && cmake .. -GNinja -DNE_BUILD_TESTS=ON -DNE_PROFILING=ON -DCMAKE_BUILD_TYPE=Release && ninja run_llama && env ENGINE_PROFILING=1 numactl -m 1 -C 56-111 bin/run_llama -m /home_local/dingyi/data/llama-7b-hf-pr347-q4j-sym-int8-fp32-g128.bin --seed 1234 -t 56 -b 2048 -c 2048 -n 1 -p "$(echo "$LUOYU_PROMPT" | cut -d' ' -f 1-1500)"
rm -rf bin && cmake .. -GNinja -DNE_BUILD_TESTS=OFF -DNE_PROFILING=OFF -DCMAKE_BUILD_TYPE=Release && ninja run_llama && env --unset ENGINE_PROFILING numactl -m 1 -C 56-111 bin/run_llama -m /home_local/dingyi/data/llama-7b-hf-pr347-q4j-sym-int8-fp32-g128.bin --seed 1234 -t 56 -b 2048 -c 2048 -n 32 -p "$GIRL_PROMPT"
```
</details>

### Main branch 
- <details>
  <summary>
  <code>env ENGINE_PROFILING=1 numactl -m 1 -C 56-111 bin/run_gptj -m /home_local/dingyi/data/gpt-j-6B-pr347-q4j-sym-int8-fp32-g128.bin --seed 1234 -t 48 -b 2048 -c 2048 -n 1 -p "$LUOYU_PROMPT"</code>
  </summary>

  ```
  === GRAPH Profiling ===
  perf_total_per_op_us[                     ADD] =  55.135 ms
  perf_total_per_op_us[                     MUL] =  15.932 ms
  perf_total_per_op_us[                    NORM] =  25.274 ms
  perf_total_per_op_us[                 MUL_MAT] =  81.835 ms
  perf_total_per_op_us[       MUL_MAT_WITH_BIAS] =  59.493 ms
  perf_total_per_op_us[                 RESHAPE] =   0.285 ms
  perf_total_per_op_us[                    VIEW] =   0.750 ms
  perf_total_per_op_us[                 PERMUTE] =   0.115 ms
  perf_total_per_op_us[                GET_ROWS] =   7.722 ms
  perf_total_per_op_us[                    ROPE] =  30.034 ms
  perf_total_per_op_us[                 MUL_QKV] = 221.504 ms
  perf_total_per_op_us[            FFN_ADD_GeLU] = 600.671 ms
  perf_total_per_op_us[              FLASH_ATTN] = 209.671 ms
  perf_total_per_op_us[    FLASH_ATTN_KV_UPDATE] =  94.790 ms
  perf_total_per_op_us[           INNER PRODUCT] =   0.000 ms
  ========================================
  ```
  </details>
- <details>
  <summary>
  <code>env ENGINE_PROFILING=1 numactl -m 1 -C 56-111 bin/run_gptj -m /home_local/dingyi/data/gpt-j-6B-pr347-q4j-sym-int8-fp32-g128.bin --seed 1234 -t 56 -b 2048 -c 2048 -n 1 -p "$LUOYU_PROMPT"</code>
  </summary>

  ```
  === GRAPH Profiling ===
  perf_total_per_op_us[                     ADD] =  43.585 ms
  perf_total_per_op_us[                     MUL] =  11.440 ms
  perf_total_per_op_us[                    NORM] =  22.630 ms
  perf_total_per_op_us[                 MUL_MAT] =  72.616 ms
  perf_total_per_op_us[       MUL_MAT_WITH_BIAS] =  63.080 ms
  perf_total_per_op_us[                 RESHAPE] =   0.320 ms
  perf_total_per_op_us[                    VIEW] =   0.709 ms
  perf_total_per_op_us[                 PERMUTE] =   0.098 ms
  perf_total_per_op_us[                GET_ROWS] =   8.266 ms
  perf_total_per_op_us[                    ROPE] =  33.194 ms
  perf_total_per_op_us[                 MUL_QKV] = 207.260 ms
  perf_total_per_op_us[            FFN_ADD_GeLU] = 584.771 ms
  perf_total_per_op_us[              FLASH_ATTN] = 194.659 ms
  perf_total_per_op_us[    FLASH_ATTN_KV_UPDATE] =  90.783 ms
  perf_total_per_op_us[           INNER PRODUCT] =   0.000 ms
  ========================================
  ```
  </details>
- <details>
  <summary>
  <code>env --unset ENGINE_PROFILING numactl -m 1 -C 56-111 bin/run_gptj -m /home_local/dingyi/data/gpt-j-6B-pr347-q4j-sym-int8-fp32-g128.bin --seed 1234 -t 56 -b 2048 -c 2048 -n 32 -p "$GIRL_PROMPT"</code>
  </summary>

  ```
  Once upon a time, there existed a little girl, who liked to have adventures. She wanted to go to places and meet new people, and have fun. This girl went by the name of LittleRed Riding Hood.
  One day Little Red Riding Hood was walking through the forest when she noticed someone following her. She
  model_print_timings:        load time =  1083.34 ms
  model_print_timings:      sample time =    28.25 ms /    32 runs   (    0.88 ms per token)
  model_print_timings: prompt eval time =    89.23 ms /    32 tokens (    2.79 ms per token)
  model_print_timings:        eval time =   716.09 ms /    31 runs   (   23.10 ms per token)
  model_print_timings:       total time =  1842.00 ms
  ========== eval time log of each prediction ==========
  prediction   0, time: 89.23ms
  prediction   1, time: 28.39ms
  prediction   2, time: 26.08ms
  prediction   3, time: 23.91ms
  prediction   4, time: 23.78ms
  prediction   5, time: 24.07ms
  prediction   6, time: 24.24ms
  prediction   7, time: 23.89ms
  prediction   8, time: 23.18ms
  prediction   9, time: 22.95ms
  prediction  10, time: 22.72ms
  prediction  11, time: 22.56ms
  prediction  12, time: 22.70ms
  prediction  13, time: 22.48ms
  prediction  14, time: 22.42ms
  prediction  15, time: 22.27ms
  prediction  16, time: 22.65ms
  prediction  17, time: 22.93ms
  prediction  18, time: 22.40ms
  prediction  19, time: 22.64ms
  prediction  20, time: 22.42ms
  prediction  21, time: 22.71ms
  prediction  22, time: 22.46ms
  prediction  23, time: 22.66ms
  prediction  24, time: 22.50ms
  prediction  25, time: 22.40ms
  prediction  26, time: 22.33ms
  prediction  27, time: 22.53ms
  prediction  28, time: 22.52ms
  prediction  29, time: 22.74ms
  prediction  30, time: 22.16ms
  prediction  31, time: 22.38ms
  ```
  </details>
- <details>
  <summary>
  <code>env ENGINE_PROFILING=1 numactl -m 1 -C 56-111 bin/run_llama -m /home_local/dingyi/data/llama-7b-hf-pr347-q4j-sym-int8-fp32-g128.bin --seed 1234 -t 56 -b 2048 -c 2048 -n 1 -p "$(echo "$LUOYU_PROMPT" | cut -d' ' -f 1-1500)"</code>
  </summary>

  ```
  === GRAPH Profiling ===
  perf_total_per_op_us[                     ADD] =  61.765 ms
  perf_total_per_op_us[                     MUL] =  40.942 ms
  perf_total_per_op_us[                RMS_NORM] =  60.443 ms
  perf_total_per_op_us[                 MUL_MAT] = 149.868 ms
  perf_total_per_op_us[                 RESHAPE] =   0.450 ms
  perf_total_per_op_us[                    VIEW] =   0.980 ms
  perf_total_per_op_us[                 PERMUTE] =   0.115 ms
  perf_total_per_op_us[               TRANSPOSE] =   0.103 ms
  perf_total_per_op_us[                GET_ROWS] =   7.885 ms
  perf_total_per_op_us[                    ROPE] =  50.016 ms
  perf_total_per_op_us[                 MUL_QKV] = 285.974 ms
  perf_total_per_op_us[                FFN_SILU] = 769.070 ms
  perf_total_per_op_us[              FLASH_ATTN] = 223.920 ms
  perf_total_per_op_us[    FLASH_ATTN_KV_UPDATE] = 100.957 ms
  perf_total_per_op_us[           INNER PRODUCT] =   0.000 ms
  ========================================
  ```
  </details>
- <details>
  <summary>
  <code>env --unset ENGINE_PROFILING numactl -m 1 -C 56-111 bin/run_llama -m /home_local/dingyi/data/llama-7b-hf-pr347-q4j-sym-int8-fp32-g128.bin --seed 1234 -t 56 -b 2048 -c 2048 -n 32 -p "$GIRL_PROMPT"</code>
  </summary>

  ```
  Once upon a time, there existed a little girl, who liked to have adventures. She wanted to go to places and meet new people, and have fun.
  Then one day, that little girl grew up. And she became a woman. A woman with a job, who travelled overseas for work,
  model_print_timings:        load time =  1267.53 ms
  model_print_timings:      sample time =    18.97 ms /    32 runs   (    0.59 ms per token)
  model_print_timings: prompt eval time =   105.54 ms /    34 tokens (    3.10 ms per token)
  model_print_timings:        eval time =   783.36 ms /    31 runs   (   25.27 ms per token)
  model_print_timings:       total time =  2079.16 ms
  ========== eval time log of each prediction ==========
  prediction   0, time: 105.54ms
  prediction   1, time: 26.41ms
  prediction   2, time: 25.19ms
  prediction   3, time: 25.41ms
  prediction   4, time: 25.37ms
  prediction   5, time: 25.21ms
  prediction   6, time: 25.17ms
  prediction   7, time: 25.18ms
  prediction   8, time: 25.28ms
  prediction   9, time: 25.41ms
  prediction  10, time: 25.21ms
  prediction  11, time: 25.12ms
  prediction  12, time: 25.15ms
  prediction  13, time: 25.11ms
  prediction  14, time: 25.07ms
  prediction  15, time: 25.54ms
  prediction  16, time: 25.32ms
  prediction  17, time: 24.97ms
  prediction  18, time: 25.49ms
  prediction  19, time: 25.07ms
  prediction  20, time: 25.04ms
  prediction  21, time: 25.07ms
  prediction  22, time: 25.33ms
  prediction  23, time: 25.16ms
  prediction  24, time: 25.20ms
  prediction  25, time: 25.42ms
  prediction  26, time: 25.51ms
  prediction  27, time: 25.27ms
  prediction  28, time: 25.41ms
  prediction  29, time: 25.02ms
  prediction  30, time: 25.28ms
  prediction  31, time: 25.00ms
  ```
  </details>

### This branch
- <details>
  <summary>
  <code>env ENGINE_PROFILING=1 numactl -m 1 -C 56-111 bin/run_gptj -m /home_local/dingyi/data/gpt-j-6B-pr347-q4j-sym-int8-fp32-g128.bin --seed 1234 -t 48 -b 2048 -c 2048 -n 1 -p "$LUOYU_PROMPT"</code>
  </summary>

  ```
  === GRAPH Profiling ===
  perf_total_per_op_us[                     ADD] =  49.759 ms
  perf_total_per_op_us[                     MUL] =  14.224 ms
  perf_total_per_op_us[                    NORM] =  22.228 ms
  perf_total_per_op_us[                 MUL_MAT] =  81.711 ms
  perf_total_per_op_us[       MUL_MAT_WITH_BIAS] =  59.821 ms
  perf_total_per_op_us[                 RESHAPE] =   0.300 ms
  perf_total_per_op_us[                    VIEW] =   0.700 ms
  perf_total_per_op_us[                 PERMUTE] =   0.093 ms
  perf_total_per_op_us[                GET_ROWS] =   7.708 ms
  perf_total_per_op_us[                    ROPE] =  30.821 ms
  perf_total_per_op_us[                 MUL_QKV] = 226.942 ms
  perf_total_per_op_us[            FFN_ADD_GeLU] = 599.899 ms
  perf_total_per_op_us[              FLASH_ATTN] = 146.544 ms
  perf_total_per_op_us[    FLASH_ATTN_KV_UPDATE] =  56.004 ms
  perf_total_per_op_us[           INNER PRODUCT] =   0.000 ms
  ========================================
  ```
  </details>
- <details>
  <summary>
  <code>env ENGINE_PROFILING=1 numactl -m 1 -C 56-111 bin/run_gptj -m /home_local/dingyi/data/gpt-j-6B-pr347-q4j-sym-int8-fp32-g128.bin --seed 1234 -t 56 -b 2048 -c 2048 -n 1 -p "$LUOYU_PROMPT"</code>
  </summary>

  ```
  === GRAPH Profiling ===
  perf_total_per_op_us[                     ADD] =  44.710 ms
  perf_total_per_op_us[                     MUL] =  11.468 ms
  perf_total_per_op_us[                    NORM] =  21.851 ms
  perf_total_per_op_us[                 MUL_MAT] =  79.625 ms
  perf_total_per_op_us[       MUL_MAT_WITH_BIAS] =  66.223 ms
  perf_total_per_op_us[                 RESHAPE] =   0.334 ms
  perf_total_per_op_us[                    VIEW] =   0.708 ms
  perf_total_per_op_us[                 PERMUTE] =   0.118 ms
  perf_total_per_op_us[                GET_ROWS] =   8.470 ms
  perf_total_per_op_us[                    ROPE] =  42.760 ms
  perf_total_per_op_us[                 MUL_QKV] = 215.438 ms
  perf_total_per_op_us[            FFN_ADD_GeLU] = 592.271 ms
  perf_total_per_op_us[              FLASH_ATTN] = 154.700 ms
  perf_total_per_op_us[    FLASH_ATTN_KV_UPDATE] =  58.654 ms
  perf_total_per_op_us[           INNER PRODUCT] =   0.000 ms
  ========================================
  ```
  </details>
- <details>
  <summary>
  <code>env --unset ENGINE_PROFILING numactl -m 1 -C 56-111 bin/run_gptj -m /home_local/dingyi/data/gpt-j-6B-pr347-q4j-sym-int8-fp32-g128.bin --seed 1234 -t 56 -b 2048 -c 2048 -n 32 -p "$GIRL_PROMPT"</code>
  </summary>

  ```
  Once upon a time, there existed a little girl, who liked to have adventures. She wanted to go to places and meet new people, and have fun. This girl went by the name of LittleRed Riding Hood.
  One day Little Red Riding Hood was walking through the forest when she noticed someone following her. She
  model_print_timings:        load time =  1040.85 ms
  model_print_timings:      sample time =    32.42 ms /    32 runs   (    1.01 ms per token)
  model_print_timings: prompt eval time =    88.35 ms /    32 tokens (    2.76 ms per token)
  model_print_timings:        eval time =   724.59 ms /    31 runs   (   23.37 ms per token)
  model_print_timings:       total time =  1812.39 ms
  ========== eval time log of each prediction ==========
  prediction   0, time: 88.35ms
  prediction   1, time: 26.03ms
  prediction   2, time: 24.79ms
  prediction   3, time: 24.79ms
  prediction   4, time: 25.17ms
  prediction   5, time: 25.08ms
  prediction   6, time: 24.60ms
  prediction   7, time: 23.21ms
  prediction   8, time: 22.98ms
  prediction   9, time: 23.08ms
  prediction  10, time: 23.43ms
  prediction  11, time: 23.05ms
  prediction  12, time: 22.79ms
  prediction  13, time: 23.11ms
  prediction  14, time: 22.85ms
  prediction  15, time: 22.78ms
  prediction  16, time: 23.10ms
  prediction  17, time: 23.08ms
  prediction  18, time: 22.82ms
  prediction  19, time: 23.02ms
  prediction  20, time: 22.91ms
  prediction  21, time: 23.00ms
  prediction  22, time: 22.64ms
  prediction  23, time: 22.92ms
  prediction  24, time: 22.91ms
  prediction  25, time: 22.67ms
  prediction  26, time: 23.11ms
  prediction  27, time: 22.85ms
  prediction  28, time: 23.03ms
  prediction  29, time: 22.81ms
  prediction  30, time: 23.05ms
  prediction  31, time: 22.96ms
  ```
  </details>
- <details>
  <summary>
  <code>env ENGINE_PROFILING=1 numactl -m 1 -C 56-111 bin/run_llama -m /home_local/dingyi/data/llama-7b-hf-pr347-q4j-sym-int8-fp32-g128.bin --seed 1234 -t 56 -b 2048 -c 2048 -n 1 -p "$(echo "$LUOYU_PROMPT" | cut -d' ' -f 1-1500)"</code>
  </summary>

  ```
  === GRAPH Profiling ===
  perf_total_per_op_us[                     ADD] =  45.444 ms
  perf_total_per_op_us[                     MUL] =  25.363 ms
  perf_total_per_op_us[                RMS_NORM] =  47.991 ms
  perf_total_per_op_us[                 MUL_MAT] = 135.003 ms
  perf_total_per_op_us[                 RESHAPE] =   0.427 ms
  perf_total_per_op_us[                    VIEW] =   0.984 ms
  perf_total_per_op_us[                 PERMUTE] =   0.108 ms
  perf_total_per_op_us[               TRANSPOSE] =   0.102 ms
  perf_total_per_op_us[                GET_ROWS] =   8.830 ms
  perf_total_per_op_us[                    ROPE] =  50.482 ms
  perf_total_per_op_us[                 MUL_QKV] = 277.234 ms
  perf_total_per_op_us[                FFN_SILU] = 707.648 ms
  perf_total_per_op_us[              FLASH_ATTN] = 187.805 ms
  perf_total_per_op_us[    FLASH_ATTN_KV_UPDATE] =  72.886 ms
  perf_total_per_op_us[           INNER PRODUCT] =   0.000 ms
  ========================================
  ```
  </details>
- <details>
  <summary>
  <code>env --unset ENGINE_PROFILING numactl -m 1 -C 56-111 bin/run_llama -m /home_local/dingyi/data/llama-7b-hf-pr347-q4j-sym-int8-fp32-g128.bin --seed 1234 -t 56 -b 2048 -c 2048 -n 32 -p "$GIRL_PROMPT"</code>
  </summary>

  ```
  Once upon a time, there existed a little girl, who liked to have adventures. She wanted to go to places and meet new people, and have fun.
  Then one day, that little girl grew up. And she became a woman. A woman with a job, who travelled overseas for work,
  model_print_timings:        load time =  1308.79 ms
  model_print_timings:      sample time =    18.67 ms /    32 runs   (    0.58 ms per token)
  model_print_timings: prompt eval time =   103.91 ms /    34 tokens (    3.06 ms per token)
  model_print_timings:        eval time =   822.86 ms /    31 runs   (   26.54 ms per token)
  model_print_timings:       total time =  2160.33 ms
  ========== eval time log of each prediction ==========
  prediction   0, time: 103.91ms
  prediction   1, time: 29.13ms
  prediction   2, time: 28.06ms
  prediction   3, time: 27.99ms
  prediction   4, time: 28.03ms
  prediction   5, time: 28.10ms
  prediction   6, time: 27.85ms
  prediction   7, time: 28.10ms
  prediction   8, time: 27.98ms
  prediction   9, time: 27.91ms
  prediction  10, time: 27.99ms
  prediction  11, time: 27.92ms
  prediction  12, time: 27.93ms
  prediction  13, time: 28.07ms
  prediction  14, time: 28.06ms
  prediction  15, time: 26.11ms
  prediction  16, time: 27.22ms
  prediction  17, time: 24.95ms
  prediction  18, time: 25.32ms
  prediction  19, time: 24.98ms
  prediction  20, time: 25.28ms
  prediction  21, time: 25.05ms
  prediction  22, time: 24.94ms
  prediction  23, time: 24.98ms
  prediction  24, time: 25.09ms
  prediction  25, time: 24.97ms
  prediction  26, time: 25.02ms
  prediction  27, time: 25.34ms
  prediction  28, time: 25.38ms
  prediction  29, time: 24.99ms
  prediction  30, time: 24.75ms
  prediction  31, time: 25.36ms
  ```
  </details>
In conclusion, this PR should brings ~40% OP level performance for first-token inference in large scale problems for models that with fused MHA enabled, while not bringing any accuracy compromise. Next token optimization will be investigated later.

## Dependency Change?
N/A

### CODE REVIEW

File: intel_extension_for_transformers/llm/runtime/graph/models/model_utils/model_utils.cpp [hunk 1]
```diff
@@ -81,9 +81,11 @@ static bool kv_cache_init(const struct model_hparams& hparams, struct model_kv_c
   }
 
   // NE_TYPE_JBLAS can not be allocated memory
-  cache.k = ne_new_tensor_1d(cache.ctx, wtype == NE_TYPE_JBLAS ? NE_TYPE_I8 : wtype, n_elements_k, NE_SIZE_CALC);
+  cache.k = ne_new_tensor_1d(cache.ctx, wtype == NE_TYPE_JBLAS ? NE_TYPE_I8 : wtype, n_elements_k + 64, NE_SIZE_CALC);
```
Comment (airMeng) [inline intel_extension_for_transformers/llm/runtime/graph/models/model_utils/model_utils.cpp:84]: make 64 as an static variable or get from some static function

File: intel_extension_for_transformers/llm/runtime/graph/core/layers/mha_dense.cpp [hunk 2]
```diff
@@ -1856,52 +1857,73 @@ void jblas_reordered_attn_fp32_update_k(const jblas_fusion_attn_fp32_update_kv_a
...
   NE_ASSERT(p.step_head_size == 1);
   const auto pad_headsize = padto(p.head_size, 48);
   const auto pad_seq_max = padto(p.seq_max, 32);
   const auto step_cache_head_num = pad_headsize * pad_seq_max;
   const auto step_cache_bs = p.head_num * step_cache_head_num;
+  GetCPUDevice();
+  const bool use_jit = _cd->AVX512_BF16() && (p.seq_off == 0);
 
+  jblas::utils::timer<jblas::utils::microseconds> tm;
+  tm.start();
```
Comment (airMeng) [inline intel_extension_for_transformers/llm/runtime/graph/core/layers/mha_dense.cpp:1899]: for debugging?

File: intel_extension_for_transformers/llm/library/jblas/jblas/kernel_jit.h [hunk 3]
```diff
@@ -740,6 +743,430 @@ static inline JBLAS_CODE decompress_s4_s8(utils::int4x2* srcptr, int8_t* dstptr,
...
+    lea(reg_tmp, ptr[reg_dstptr + reg_tmp1]);
+    lea(reg_tmp1, ptr[reg_srcptr + reg_itercol * src_bytes]);
+    for (int jj = 0; jj < n_tile; jj += ne_zmm) {
+      generate_Nbitsmask(mask_rd, reg_itercol, ptr[reg_colsize - jj], reg_tmp2, reg_tmp3, ne_zmm);
+      for (int ii = 0; ii < row_pack; ii++) {
+        const Xbyak::Xmm reg_srcs_ii = src_bytes_vmm == 64   ? Xbyak::Zmm(reg_srcs[ii].getIdx())
+                                       : src_bytes_vmm == 32 ? Xbyak::Ymm(reg_srcs[ii].getIdx())
+                                       : src_bytes_vmm == 16 ? Xbyak::Xmm(reg_srcs[ii].getIdx())
+                                                             : (assert(false), reg_srcs[ii]);
+        if (src_bytes == 1) {
```
Comment (airMeng) [inline intel_extension_for_transformers/llm/library/jblas/jblas/kernel_jit.h:848]: constexpr

File: intel_extension_for_transformers/llm/library/jblas/jblas/kernel_jit.h [hunk 4]
```diff
@@ -740,6 +743,430 @@ static inline JBLAS_CODE decompress_s4_s8(utils::int4x2* srcptr, int8_t* dstptr,
...
+
+    L(".tailrowloop");
+    L(".tailcolloop");
+    mov(reg_tmp1, reg_itercol);
+    imul(reg_tmp1, reg_dststride);
+    lea(reg_tmp, ptr[reg_dstptr + reg_tmp1]);
+    lea(reg_tmp1, ptr[reg_srcptr + reg_itercol * src_bytes]);
+    for (int jj = 0; jj < n_tile; jj += ne_zmm) {
+      generate_Nbitsmask(mask_rd, reg_itercol, ptr[reg_colsize - jj], reg_tmp2, reg_tmp3, ne_zmm);
+      if (row_pack == 2) {
```
Comment (airMeng) [inline intel_extension_for_transformers/llm/library/jblas/jblas/kernel_jit.h:884]: constexpr

File: intel_extension_for_transformers/llm/library/jblas/jblas/kernel_jit.h [hunk 5]
```diff
@@ -740,6 +743,430 @@ static inline JBLAS_CODE decompress_s4_s8(utils::int4x2* srcptr, int8_t* dstptr,
...
+  PaddingInterleaveCvt(int n_tile, JBLAS_DTYPE dst_t, JBLAS_DTYPE src_t, int row_pack = 0) : xbyak::JitAvx512f() {
+    inLocalLabel();  // use local label for multiple instance
+    const auto src_bytes = static_cast<int>(utils::jblas_dtype_size(src_t));
+    const auto dst_bytes = static_cast<int>(utils::jblas_dtype_size(dst_t));
+    if (row_pack == 0) row_pack = 4 / dst_bytes;  // default value
+    const auto ne_zmm = 64 / std::max(src_bytes, dst_bytes);
+    const auto src_bytes_vmm = ne_zmm * src_bytes;
+    const auto dst_bytes_vmm = ne_zmm * dst_bytes;
+
+    assert(n_tile % ne_zmm == 0);
```
Comment (airMeng) [inline intel_extension_for_transformers/llm/library/jblas/jblas/kernel_jit.h:778]: should be ```static_assert```?
throw some comments about why fail

File: intel_extension_for_transformers/llm/library/jblas/jblas/kernel_jit.h (hunk 4 above)
Comment (DDEle) [reply intel_extension_for_transformers/llm/library/jblas/jblas/kernel_jit.h:884]: I tried to avoid the use of template in code generators to potentially save some time for compilation. It should be harmless to add some runtime branches in the logic of JIT code generation. As a result, it is not a constexpr here.

File: intel_extension_for_transformers/llm/library/jblas/jblas/kernel_jit.h (hunk 5 above)
Comment (DDEle) [reply intel_extension_for_transformers/llm/library/jblas/jblas/kernel_jit.h:778]: I tried to avoid the use of template in code generators to potentially save some time for compilation. It should be harmless to add some runtime branches in the logic of JIT code generation. As a result, it is not a constexpr here.

File: intel_extension_for_transformers/llm/library/jblas/jblas/kernel_jit.h (hunk 3 above)
Comment (DDEle) [reply intel_extension_for_transformers/llm/library/jblas/jblas/kernel_jit.h:848]: I tried to avoid the use of template in code generators to potentially save some time for compilation. It should be harmless to add some runtime branches in the logic of JIT code generation. As a result, it is not a constexpr here.

**Comment by airMeng** [comment]
can you try more models to verify the performance gain?

File: intel_extension_for_transformers/llm/runtime/graph/core/layers/mha_dense.cpp (hunk 2 above)
Comment (DDEle) [reply intel_extension_for_transformers/llm/runtime/graph/core/layers/mha_dense.cpp]: Removed in 79acbd5

File: intel_extension_for_transformers/llm/runtime/graph/models/model_utils/model_utils.cpp (hunk 1 above)
Comment (DDEle) [reply intel_extension_for_transformers/llm/runtime/graph/models/model_utils/model_utils.cpp]: Changed in 79acbd5. (Using name inspired from `intel_extension_for_transformers/llm/runtime/deprecated/executor/include/i_malloc.hpp`)

File: intel_extension_for_transformers/llm/runtime/graph/models/model_utils/model_utils.cpp [hunk 6]
```diff
@@ -81,9 +81,14 @@ static bool kv_cache_init(const struct model_hparams& hparams, struct model_kv_c
   }
 
   // NE_TYPE_JBLAS can not be allocated memory
-  cache.k = ne_new_tensor_1d(cache.ctx, wtype == NE_TYPE_JBLAS ? NE_TYPE_I8 : wtype, n_elements_k, NE_SIZE_CALC);
+  const auto wtype_alloc = wtype == NE_TYPE_JBLAS ? NE_TYPE_I8;
+  cache.k = ne_new_tensor_1d(cache.ctx, wtype_alloc, n_elements_k + NE_ALIGNMENT, NE_SIZE_CALC);
```
Comment (airMeng) [inline intel_extension_for_transformers/llm/runtime/graph/models/model_utils/model_utils.cpp:85]: if 64 is what we need to align to, should it be ```(n_elements_k - NE_ALIGNMENT)/NE_ALIGNMENT*NE_ALIGNMENT```

Comment (DDEle) [reply intel_extension_for_transformers/llm/runtime/graph/models/model_utils/model_utils.cpp:85]: Well, currently there is no guarantee that it will provides aligned memory chunks. Therefore, I ask a bit more during allocation and applied an offset (ne_view_1d) to get a tensor with aligned memory.

File: intel_extension_for_transformers/llm/library/jblas/jblas/kernel_jit.h (hunk 5 above)
Comment (airMeng) [reply intel_extension_for_transformers/llm/library/jblas/jblas/kernel_jit.h:778]: ok, for these asserts, please throw some explanation, just for example
```suggestion
    assert(n_tile % ne_zmm == 0, "n_tile must be divided by (64/sizeof(dt))");
```

Comment (DDEle) [reply intel_extension_for_transformers/llm/library/jblas/jblas/kernel_jit.h:778]: Unlike `static_assert`, assert marco function [does not have a second parameter](https://en.cppreference.com/w/cpp/error/assert). But yes, we can try to use the workaround introduced in the link.

However, I would like to leave it as it is for now until next time when I need to update jblas to avoid jblas syncing conflictions.  (and I'll remember to invite you to review).

**Review by airMeng — APPROVED** [review]

**Comment by DDEle** [comment]
> can you try more models to verify the performance gain?

This improvement will only benefits models with fused MHA enabled (i.e. gptj & llama only by now).

I have added some test result for llama in the PR description. Looks that the result is similar with that of gptj.

**Review by luoyu-intel — APPROVED** [review]

**Comment by VincyZhang** [comment]
ext test: https://inteltf-jenk.sh.intel.com/view/nlp-toolkit-validation/job/ITREX-cpp-graph-extension/97/

**Comment by VincyZhang** [comment]
https://inteltf-jenk.sh.intel.com/job/ITREX-cpp-graph-extension/99/
